### ISSUE TITLE
add early stopping on training loss

### ISSUE THREAD

**Issue by AlicjaA** [AlicjaA/zebras#4]

**Comment by AlxndrMlk** [comment]
Hey @WMSaj13 

any progress on this issue?

**Comment by WMSaj13** [comment]
@AlxndrMlk sorry for the delays due to work, and need for the rest,  

1. I will code and commit this week at least simple early stopping on given (smoothed exponentially) loss can be implemented with ease and generally in https://github.com/AlicjaA/zebras/blob/master/train.py (e.g. line 65 is an explicit fetch of loss values) so I think of just adding 

in the same file:
- storage of previous values losses values in the function for smoothing
- smooth function
- a few lines of logic calculating smoothed losses and stoping

and in the https://github.com/AlicjaA/zebras/blob/master/options/train_options.py
- extra input param to train.py, path to json file storing the name of the loss(es) name(es) with it stop values plus smoothing params maybe some boolean switch if all or any of criteria should be fullfiled

2. as for the more developed criteria of training stop I think that we should consider  some other metrics like  Fréchet Inception Distance 
- figure 2 in https://www.researchgate.net/publication/341462013_Deep_Snow_Synthesizing_Remote_Sensing_Imagery_with_Generative_Adversarial_Nets/figures
- https://github.com/hukkelas/pytorch-frechet-inception-distance
- https://machinelearningmastery.com/how-to-implement-the-frechet-inception-distance-fid-from-scratch/

**Comment by AlxndrMlk** [comment]
No worries @WMSaj13 ,

I could say the same. It was a very busy time for me as well. 

Thanks for sharing the ideas! 💡

> in the same file:
> 
>     storage of previous values losses values in the function for smoothing
>     smooth function
>     a few lines of logic calculating smoothed losses and stoping


I was wondering if we need smoothing at all. Maybe a simple early stopping policy with patience would be enough? What are your thoughts?


>     as for the more developed criteria of training stop I think that we should consider some other metrics like Fréchet Inception Distance
> 
>     figure 2 in https://www.researchgate.net/publication/341462013_Deep_Snow_Synthesizing_Remote_Sensing_Imagery_with_Generative_Adversarial_Nets/figures
>     https://github.com/hukkelas/pytorch-frechet-inception-distance
>     https://machinelearningmastery.com/how-to-implement-the-frechet-inception-distance-fid-from-scratch/

Sounds interesting! Will try to find a while later this week to dig deeper into this topic.

**Comment by WMSaj13** [comment]
> I was wondering if we need smoothing at all. Maybe a simple early stopping policy with patience would be enough? What are your thoughts?
Yes, I think we can stop if change is within given bounds for given period of time ( and leave smoothing for plots to get insight)

**Comment by AlxndrMlk** [comment]
I like this idea!

**Comment by WMSaj13** [comment]
@AlicjaA , @AlxndrMlk   

the first version of early stop on losses in fork from orginal repo (should merge without problems):
https://github.com/WMSaj13/pytorch-CycleGAN-and-pix2pix/commit/05c3af0c6dde716c5a654df3af68cb5c881d9619

Usage by additional params to train.py , 

_--es_conditions json_ : json with names of loss :  max change or pair, range of values  
_--es_mem int_   how many previous values of losses are checked
_--es_all_  stops only if all conditions apply
_--es_iter int_  how many iterations between checks (if below zero check after each epoch)  

examples :

_/workspace/pytorch-CycleGAN-and-pix2pix# python train.py --dataroot ./datasets/horse2zebra --name horse2zebra_cyclegan --model cycle_gan --pool_size 50 --no_dropout --es_conditions '{"cycle_A":dA}' --es_mem 3_ checks after each epoch if loss didnt' change more than dA ( (max- min) / 2 < dA)  stops if true   


_/workspace/pytorch-CycleGAN-and-pix2pix# python train.py --dataroot ./datasets/horse2zebra --name horse2zebra_cyclegan --model cycle_gan --pool_size 50 --no_dropout --es_conditions '{"cycle_A":[A,B],"cycle_B":[C,D]}' --es_mem 3 --es_all --es_iter 100_  checks each 100 iters if in last three times cycle_A was in range A,B and cycle_B was in range C,D then stops if both true
